editor/src/clj/editor/model_scene.clj

```diff
@@ -310,86 +256,173 @@
   (:bones content))
 
 (g/defnk produce-content [_node-id resource]
-      (model-loader/load-scene _node-id resource))
+  (model-loader/load-scene _node-id resource))
 
 (g/defnk produce-animation-info [resource]
   [{:path (resource/proj-path resource) :parent-id "" :resource resource}])
 
 (g/defnk produce-animation-ids [content]
   (:animation-ids content))
 
+(def ^:private default-material-ids ["default"])
+
 (g/defnk produce-material-ids [content]
   (let [ret (:material-ids content)]
     (if (zero? (count ret))
-      ["default"]
+      default-material-ids
       ret)))
 
-(defn- index-oob [vs is comp-count]
-  (> (* comp-count (reduce max 0 is)) (count vs)))
-
-(defn- validate-meshes [meshes]
-  (when-let [es (seq (keep (fn [m]
-                             (let [{:keys [^floats positions ^floats normals ^floats texcoord0 ^ints indices]} m]
-                               (when (or (and (not= (alength normals) 0) (not= (alength positions) (alength normals))) ; normals optional
-                                         (index-oob positions indices 3))
-                                 (error-values/error-fatal "Failed to produce vertex buffers from mesh set. The scene might contain invalid data."))))
-                       meshes))]
-    (error-values/error-aggregate es)))
-
-(g/defnk produce-scene [_node-id aabb meshes]
-  (or (validate-meshes meshes)
-      {:node-id _node-id
-       :aabb aabb
-       :renderable {:render-fn render-scene
-                    :tags #{:model}
-                    :batch-key _node-id
-                    :select-batch-key _node-id
-                    :user-data {:meshes meshes
-                                :shader shader-pos-nrm-tex
-                                :textures {"texture" @texture/white-pixel}}
-                    :passes [pass/opaque pass/opaque-selection]}
-       :children [{:node-id _node-id
-                   :aabb aabb
-                   :renderable {:render-fn render-outline
-                                :tags #{:model :outline}
-                                :batch-key _node-id
-                                :select-batch-key _node-id
-                                :passes [pass/outline]}}]}))
+(defn- make-renderable-mesh [mesh mesh-material-index->material-name]
+  (let [mesh-renderable-data (mesh->renderable-data mesh)]
+    (if (g/error-value? mesh-renderable-data)
+      mesh-renderable-data
+      (let [{:keys [aabb-min aabb-max ^int material-index]} mesh
+            mesh-aabb (geom/coords->aabb aabb-min aabb-max)
+            material-name (mesh-material-index->material-name material-index)]
+        {:aabb mesh-aabb
+         :material-name material-name
+         :renderable-data mesh-renderable-data}))))
```

Comment:
We do less work in the `render-fn` now. We can prepare the `renderable-datas` and cache them in the `renderable-mesh-set` output before producing the scene.

Potentially we might be able to fully create and cache the populated `VertexBuffers` here, but currently we rely on getting the `shader-bound-attributes` from the OpenGL context that is only available to us in the `render-fn`. I think we may want to introduce a shared "resource" OpenGL context that we use to store vertex buffers and textures so they can be shared between scene views, but that is a larger change that could have potential issues with some hardware.